File: 0210-course-schedule-ii/0210-course-schedule-ii.cpp
class Solution {
public:
    vector<int> findOrder(int numCourses, vector<vector<int>>& prerequisites) {
        vector<vector<int>> course_order(numCourses);
        vector<int> in_degree(numCourses);

        for (auto& courses: prerequisites) {
            // b is a's prerequisite
            course_order[courses[1]].push_back(courses[0]);
            in_degree[courses[0]]++;
        }

        queue<int> q;
        for (int i = 0; i < numCourses; i++) {
            if (in_degree[i] == 0) q.push(i);
        }
        
        vector<int> results;

        while (!q.empty()) {
            int current_course = q.front();
            q.pop();
            results.push_back(current_course);
            
            if (results.size() == numCourses) break;

            for (int next_course: course_order[current_course]) {
                // has been processed or pushed to q
                if (in_degree[next_course] == 0) continue;
                in_degree[next_course]--;
                if (in_degree[next_course] == 0) {
                    q.push(next_course);
                }
            }
        }

        if (results.size() == numCourses) {
            return results;
        }
        return {};
        // return results.size() == numCourses ? results : {};
    }
};

File: 0210-course-schedule-ii/0210-course-schedule-ii.py
class Solution:
    def findOrder(self, numCourses: int, prerequisites: List[List[int]]) -> List[int]:
        # build DAG and indegree
        graph = [[] for _ in range(numCourses)]
        indegree = [0] * numCourses
        for course, pre in prerequisites:
            graph[pre].append(course)
            indegree[course] += 1

        # gather those with indegree 0
        q = deque(i for i, v in enumerate(indegree) if v == 0)

        # multi-source bfs
        order = []

        while q:
            curr_course = q.popleft()
            order.append(curr_course)
            for next_course in graph[curr_course]:
                indegree[next_course] -= 1
                if indegree[next_course] == 0:
                    q.append(next_course)

        return order if len(order) == numCourses else []